pkg/recommender/recommender.go

```diff
@@ -0,0 +1,129 @@
+	p, err := metrics.NewPrometheusProvider(ctx, c, logger, metricsSelector)
+	if err != nil {
+		return nil, errors.Wrap(err, "create metrics provider")
+	}
+
+	return &recommender{
+		client:          c,
+		logger:          logger,
+		metricsProvider: p,
+	}, nil
+}
+
+func (r *recommender) RunOnce(ctx context.Context) error {
+	scas := &v1alpha1.ScyllaClusterAutoscalerList{}
+	err := r.client.List(ctx, scas)
+	if err != nil {
+		return errors.Wrap(err, "fetch SCAs")
+	}
+
+	for idx, _ := range scas.Items {
+		sca := &scas.Items[idx]
+		targetRef := sca.Spec.TargetRef
+
+		sc := &scyllav1.ScyllaCluster{}
+		err = r.client.Get(ctx, client.ObjectKey{
+			Namespace: targetRef.Namespace,
+			Name:      targetRef.Name,
+		}, sc)
+		if err != nil {
+			r.logger.Error(ctx, "fetch referenced ScyllaCluster", "error", err)
+			continue
+		}
+
+		ruleGroups := sca.Spec.ScalingRuleGroups
+		if len(sca.Spec.ScalingPolicy.Datacenters) == 0 {
+			continue
+		}
+
+		dc := sc.Spec.Datacenter
+		rackRecommendations := make([]v1alpha1.RackRecommendations, 0)
+		for _, rackScalingPolicy := range sca.Spec.ScalingPolicy.Datacenters[0].RackScalingPolicies {
+			rackName := rackScalingPolicy.Name
+			i := 0
+			for i := range dc.Racks {
+				if dc.Racks[i].Name == rackName {
+					break
+				}
+			}
+			if i == len(dc.Racks) {
+				r.logger.Error(ctx, "find rack", "error", errors.New("rack not found"), "rack", rackName)
+				continue
+			}
+			rack := dc.Racks[i]
+
+			group, found := ruleGroups[rackScalingPolicy.ScalingRuleGroup]
+			if !found {
+				r.logger.Error(ctx, "get rule group", "error", errors.New("not found"))
+				continue
+			}
+
+			recommendedMembers := rack.Members
+			for _, rule := range group.Rules {
+				labelsMap := rule.Labels
+				labelsMap["scylla_cluster"] = sc.Name
+				labelsMap["scylla_datacenter"] = dc.Name
+				labelsMap["scylla_rack"] = rackScalingPolicy.Name
+
+				res, err := r.metricsProvider.FetchMetric(ctx, rule.Metric, labelsMap)
+				if err != nil {
+					r.logger.Error(ctx, "fetch rack metrics", "rack", rack.Name, "err", err)
+					continue
+				}
+
+				if res < rule.Threshold {
+					continue
+				}
+
+				newMembers := rack.Members + rule.Instances
+				if newMembers > recommendedMembers {
+					recommendedMembers = newMembers
+				}
+			}
+			rackRecommendations = append(rackRecommendations, v1alpha1.RackRecommendations{Name: rack.Name, Members: &v1alpha1.RecommendedRackMembers{Target: recommendedMembers}})
+		}
+		recommendations := &v1alpha1.ScyllaClusterRecommendations{DataCenterRecommendations: []v1alpha1.DataCenterRecommendations{{Name: dc.Name, RackRecommendations: rackRecommendations}}}
+
+		str, err := json.Marshal(recommendations)
+		r.logger.Info(ctx, "recommendations", "sca", sca.Name, "json", string(str))
+		if err != nil {
+			r.logger.Error(ctx, "marshal recommendations", "error", err)
+			continue
+		}
```

Comment:
If you would move this log inside above loop, we wouldn't have json in side json log message.